### PR TITLE
[Backport][Inputstream] Add IChapter interface #16581

### PR DESCRIPTION
**Pull request by peak3d**
## Description
This backports IChapter Interface implementation for inputstream addons (#16581) .

## Motivation and Context
There are some addons hold back because of missing support.

## How Has This Been Tested?
not yet tested

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by peak3d** [comment]
@Rechi what is with clang format for backports ?

**Comment by peak3d** [comment]
jenkins build this please

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h [hunk 1]
```diff
@@ -461,6 +472,35 @@ namespace addon
...
+  * Return number of available chapters
+  * @remarks
+  */
+  virtual int GetChapterCount() { return 0; };
+
+  /*!
+  * Return name of chapter # ch
+  * @remarks
+  */
+  virtual const char* GetChapterName(int ch) { return nullptr; };
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h]: The changes are looking good, only about here, is it possible to use `std::string`? Then it match more the C++ Interface.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h [hunk 2]
```diff
@@ -677,8 +737,32 @@ namespace addon
...
+    }
+
+    inline static int ADDON_GetChapterCount(const AddonInstance_InputStream* instance)
+    {
+      return instance->toAddon.addonInstance->GetChapterCount();
+    }
+
+    inline static const char* ADDON_GetChapterName(const AddonInstance_InputStream* instance, int ch)
+    {
+      return instance->toAddon.addonInstance->GetChapterName(ch);
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h]: And here:
```cpp
      return instance->toAddon.addonInstance->GetChapterName(ch).c_str();
```

File: xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp [hunk 3]
```diff
@@ -477,6 +478,59 @@ bool CInputStreamAddon::IsRealtime()
...
+
+int CInputStreamAddon::GetChapterCount()
+{
+  if (m_struct.toAddon.get_chapter_count)
+    return m_struct.toAddon.get_chapter_count(&m_struct);
+
+  return 0;
+}
+
+void CInputStreamAddon::GetChapterName(std::string& name, int ch)
```
Comment (AlwinEsch) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp]: Only a question. Is this `std::string& name` value always a empty string on his call? Think yes (seen this two places below where it becomes called).
Not that it become in conflict if `std::string` where never should give `nullptr` for the `if (res)` check (but this check is always good and more safe :smile:).

**Comment by dagwieers** [comment]
@peak3d This is amazing !

**Comment by peak3d** [comment]
Jenkis build this please

**Comment by peak3d** [comment]
Jenkins build this please

**Comment by dagwieers** [comment]
> **Detected "please". Jenkins voice commands activated.**

**Comment by peak3d** [comment]
@michaelarnauts do youhave the chance to test this build with your inputstream.adaptive backport version?

**Comment by michaelarnauts** [comment]
@peak3d was this the blocking issue to make a new release of inputstream.adaptive for Kodi 18 with the Period support backported?

**Comment by dagwieers** [comment]
There is a feature request open to have a python interface to add chapters to an existing stream.
https://forum.kodi.tv/showthread.php?tid=347080

**Pull request merged**